go/vt/wrangler/reparent.go

```diff
@@ -812,3 +813,65 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
 
 	return nil
 }
+
+// TabletExternallyReparented updates all topo records so the current
+// tablet is the new master for this shard.
+func (wr *Wrangler) TabletExternallyReparented(ctx context.Context, keyspace, shard string, newMasterAlias *topodatapb.TabletAlias) error {
+
+	// Check the global shard record.
+	si, err := wr.ts.GetShard(ctx, keyspace, shard)
+	if err != nil {
+		log.Warningf("TabletExternallyReparented: failed to read global shard record for %v/%v: %v", keyspace, shard, err)
+		return err
+	}
+
+	tabletInfo, err := wr.ts.GetTablet(ctx, newMasterAlias)
+	if err != nil {
+		log.Warningf("TabletExternallyReparented: failed to read tablet record for %v: %v", newMasterAlias, err)
+		return err
+	}
+	tablet := tabletInfo.Tablet
+	// Create a reusable Reparent event with available info.
+	ev := &events.Reparent{
+		ShardInfo: *si,
+		NewMaster: *tablet,
+		OldMaster: topodatapb.Tablet{
+			Alias: si.MasterAlias,
+			Type:  topodatapb.TabletType_MASTER,
+		},
+	}
+	defer func() {
+		if err != nil {
+			event.DispatchUpdate(ev, "failed: "+err.Error())
+		}
+	}()
+	event.DispatchUpdate(ev, "starting external reparent")
+
+	// We may get called on the current master multiple times in order to fix incomplete external reparents.
+	// We update the tablet here only if it is not currently master
+	if tablet.Type != topodatapb.TabletType_MASTER {
+		log.Infof("TabletExternallyReparented: executing change callback for state change to MASTER")
+
+		_, err := wr.ts.UpdateTabletFields(ctx, tablet.Alias,
```

Comment:
I just remembered why TER started out as a tabletmanager RPC instead of a wrangler method. We had a principle that no one should edit a tablet record other than the tablet itself, except in the case that we suspected the tablet is dead. It still may make sense to abandon that principle, but we should put some thought into whether we're losing some kind of synchronization that used to be ensured by the tablet action lock.

Reply:
I think we should still keep that principle and call to the tablet via an RPC which does just these 2 things: change tablet type to master and refresh its state.

Reply:
It seems safe to use `func (agent *ActionAgent) ChangeType` to do this.